File: leetcode/backtracking/39. Combination Sum/py/main.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        
        def dfs(start, comb, remain):
            if remain < 0:
                return
            if remain == 0:
                res.append(list(comb))
                return
            
            for i in range(start, len(candidates)):
                comb.append(candidates[i])
                dfs(i, comb, remain - candidates[i])
                comb.pop()
                
        dfs(0, [], target)
        
        return res

File: leetcode/backtracking/39. Combination Sum/js/main.js
// The same number may be chosen from candidates an unlimited number of times.

// Time O(N!)
// Space O(N!)
const combinationSum = (candidates, target) => {
  let res = [];

  dfs(0, [], target);

  return res;

  function dfs(start, comb, sum) {
    if (sum < 0) {
      return;
    }

    if (sum == 0) {
      res.push([...comb]);
      return;
    }

    for (let i = start; i < candidates.length; i++) {
      comb.push(candidates[i]);
      dfs(i, comb, sum - candidates[i]);
      comb.pop();
    }
  }
};
